### PR TITLE
[SWY-198] Extract shared set song numbering helpers

### PR DESCRIPTION
**Pull request by justaddcl**
## Background
- Linear ticket: https://linear.app/paranoid-android/issue/SWY-198/extract-shared-set-song-numbering-helpers
- This removes duplicated song numbering logic from set detail, organization dashboard, and duplicate-set preview rendering.
- The change centralizes section start indexes and per-song display indexes in a pure helper so visible numbering behavior stays consistent across set views.

## What's changed
@coderabbitai summary
- Added `getSetSongNumbering` with generic section/song shapes and unit coverage for empty sections, multiple sections, and mixed song counts.
- Updated dashboard, set detail, and duplicate-set preview rendering to consume numbered section results instead of recalculating indexes in render loops.
- Adjusted section card/list props and shared section prop typing so display indexes come from the helper output.

## How to test
- Open an organization dashboard and confirm each set's songs are numbered sequentially across sections.
- Open a set detail page with multiple sections and confirm song numbering is unchanged, including sections with no songs.
- Open the duplicate-set dialog and confirm the song preview uses the same numbering as the source set.

<h3>Greptile Summary</h3>

This PR extracts three near-identical inline song-numbering loops (dashboard, set detail, duplicate-set preview) into a single generic pure helper, `getSetSongNumbering`, and updates all three call sites to consume its output.

- **New helper** (`getSetSongNumbering.ts`): sorts sections by `position` via `toSorted`, builds a `Map<id → sectionStartIndex>` in one pass, then re-maps the original input array to preserve caller order — correctly handling empty sections and out-of-order inputs. The `displayIndex` formula (`sectionStartIndex + song.position`) is identical to the old inline computation.
- **Props refactoring**: `SetSectionCard` and `SetSectionList` replace the separate `section` + `sectionStartIndex` props with a single `numberedSection` prop; `SetSectionCardProps` is split into `SetSectionCardSectionProps` (section identity/length/menu) and `SetSectionCardProps` (adds `numberedSection` + `onAddSongClick`), keeping downstream type consumers (`SetSectionActionMenu`, `EditSetSectionTypeForm`) aligned.
- **Tests**: four unit cases cover empty sections, multi-section propagation, mixed song counts with an empty middle section, and out-of-order position input.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the change is a pure refactor that replaces three identical inline index loops with a single well-tested helper; all call sites produce the same numbering output as before.

The helper's formula (sectionStartIndex + song.position) exactly matches the formula used in every loop it replaces. The toSorted-then-map design correctly handles out-of-order section arrays, which the old array-index-based loops did not. Tests cover all meaningful edge cases. The props restructuring is semantically equivalent — no existing consumers lose information.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/modules/sets/utils/getSetSongNumbering.ts | New pure helper that computes 1-based sectionStartIndex and per-song displayIndex by sorting sections by position, then preserving input order in the output array. Logic is correct and handles empty sections and out-of-order inputs. |
| src/modules/sets/utils/__tests__/getSetSongNumbering.test.ts | New unit test suite covering empty sections, multi-section index propagation, mixed song counts, and out-of-order position input; fixtures follow project conventions with createSong/createSection helpers. |
| src/modules/sets/components/SetSectionCard/SetSectionCard.tsx | Replaced section + sectionStartIndex props with a single numberedSection prop; rendering now iterates numberedSongs from the helper instead of recomputing indexes inline. Type split into SetSectionCardSectionProps and SetSectionCardProps is clean. |
| src/modules/sets/components/SetSectionList/SetSectionList.tsx | Same structural change as SetSectionCard — drops section + sectionStartIndex in favour of numberedSection; no logic changes. |
| src/modules/sets/components/forms/DuplicateSetForm.tsx | Inline O(n²) index loop replaced with a single getSetSongNumbering call; SetSectionList now receives numberedSection directly. Trivial Tailwind class reorder also included. |
| src/app/[organization]/page.tsx | Dashboard map loop now uses getSetSongNumbering to obtain displayIndex per song; redundant nested index recomputation removed. Import order also cleaned up. |
| src/app/[organization]/sets/[setId]/page.tsx | Set detail page passes numberedSection to SetSectionCard instead of computing sectionStartIndex inline; removes the SetSectionWithSongs cast that was previously needed. |
| src/modules/SetListCard/components/SetSectionActionMenu/SetSectionActionMenu.tsx | Updated imported type reference from SetSectionCardProps to SetSectionCardSectionProps following the prop-type split; no behavioural change. |
| src/modules/sets/components/forms/EditSetSectionTypeForm.tsx | EditSetSectionTypeFormProps base changed from Pick<SetSectionCardProps, …> to SetSectionCardSectionProps, which exposes the same three fields; semantically equivalent. |

<h3>Flowchart</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["sections\n(readonly TSection[])"] --> B["toSorted by position"]
    B --> C["forEach: build Map\nsectionId → sectionStartIndex"]
    C --> D["sections.map (input order preserved)"]
    D --> E["lookup sectionStartIndex from Map"]
    E --> F["section.songs.map\ndisplayIndex = sectionStartIndex + song.position"]
    F --> G["NumberedSetSection[]"]
    G --> H["SetSectionCard\n(numberedSection prop)"]
    G --> I["SetSectionList\n(numberedSection prop)"]
    G --> J["Dashboard SongItem\n(displayIndex)"]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A["sections\n(readonly TSection[])"] --> B["toSorted by position"]
    B --> C["forEach: build Map\nsectionId → sectionStartIndex"]
    C --> D["sections.map (input order preserved)"]
    D --> E["lookup sectionStartIndex from Map"]
    E --> F["section.songs.map\ndisplayIndex = sectionStartIndex + song.position"]
    F --> G["NumberedSetSection[]"]
    G --> H["SetSectionCard\n(numberedSection prop)"]
    G --> I["SetSectionList\n(numberedSection prop)"]
    G --> J["Dashboard SongItem\n(displayIndex)"]
```

</a>

<sub>Reviews (2): Last reviewed commit: ["Cover unordered set section numbering"](https://github.com/justaddcl/sanbi/commit/4d91c3e81e6aa13b47119301ba8e2949bb73c92d) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=39942105)</sub>

### CODE REVIEW

**Comment by justaddcl** [comment]
#### 📝 Changed routes:


| [**/\[organization\]**](https://sanbi-37lcriyv1-justaddcls-projects.vercel.app/[organization]) | [**…nization\]/sets/\[setId\]**](https://sanbi-37lcriyv1-justaddcls-projects.vercel.app/[organization]/sets/[setId]) |
|:-:|:-:|
|<a href="https://sanbi-37lcriyv1-justaddcls-projects.vercel.app/[organization]">Screenshot of /[organization]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/I0Zc2IkajvdIScEdFL5uyIHdurNGDm7dcVQpR6B0UuMIxC0DiPyq5i11Z63pBzVjjMybHdC9sqMxIB5NqBTa9Q.png">(view full size)</a> |<a href="https://sanbi-37lcriyv1-justaddcls-projects.vercel.app/[organization]/sets/[setId]">Screenshot of /[organization]/sets/[setId]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/COZbmIlwmAIWRni0P7Cpa88Iqk39SjvJGL_mzxGCBPNs93GIEbG4vVOlZ3KDW4USFKIH5_cT53Bg2hJC-Hyvyg.png">(view full size)</a> |







Commit 4d91c3e81e6aa13b47119301ba8e2949bb73c92d (https://sanbi-37lcriyv1-justaddcls-projects.vercel.app).

File: src/modules/sets/utils/__tests__/getSetSongNumbering.test.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    expect(numbering.map((section) => section.sectionStartIndex)).toEqual([
+      1, 2, 2, 5,
+    ]);
+    expect(
+      numbering.map((section) =>
+        section.songs.map((song) => song.displayIndex),
+      ),
+    ).toEqual([[1], [], [2, 3, 4], [5]]);
+  });
+});
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/sets/utils/__tests__/getSetSongNumbering.test.ts:121]: Added coverage for unordered section input in commit 4d91c3e. The new case asserts that section start indexes are computed by ascending `position` while the returned numbering preserves the caller's input order.
